Attributes and properties (#267)
javascript-tutorial/it.javascript.info: 2-ui/1-document/06-dom-attributes-and-properties/article.md

# Attributi e proprietà

> Quando il browser carica una pagina, "legge" (in altre parole: "analizza") l'HTML e genera gli oggetti DOM. Per i nodi elemento, i più degli attributi HTML standart diventano automaticamente proprietà degli oggetti DOM.

Reviewer: ```suggestion
Quando il browser carica una pagina, "legge" (in altre parole: "analizza") l'HTML e genera gli oggetti DOM. Per i nodi elemento, i più degli attributi HTML standard diventano automaticamente proprietà degli oggetti DOM.
```

Ad esempio, se il tag è `<body id="page">`, allora l'oggetto DOM ha `body.id="page"`.

> Ma la mappatura attributo-proprietà non va di pari passo! In questo capitolo faremo attenzione a separare queste due nozioni, vedremo come lavorare con esse, quando sono la stessa cosa e quando invece sono differenti. 

Reviewer: In questo caso mi sembra molto più chiara la traduzione letterale uno a uno... Se non ho interpretato male i significato, quello che vuole dire è che non tutti gli attributi hanno la corrispettiva proprietà, o comunque non con lo stesso nome

```suggestion
Ma la mappatura attributo-proprietà non è uno a uno! In questo capitolo faremo attenzione a separare queste due nozioni, vedremo come lavorare con esse, quando sono la stessa cosa e quando invece sono differenti. 
```

## Proprietà del DOM

> Abbiamo già visto proprietà integrate del DOM. Ce ne sono molte. Ma tecnicamente nessuna di esse ci limita, e se non ci sono sufficienti, possiamo aggiungerne. 

Reviewer 2: ```suggestion
Abbiamo già visto proprietà integrate del DOM. Ce ne sono molte. Ma tecnicamente nessuna di esse ci limita, e se non sono sufficienti, possiamo aggiungerne. 
```

> Abbiamo già visto proprietà integrate del DOM. Ce ne sono molte. Ma tecnicamente nessuna di esse ci limita, e se non ce ne sono abbastanza, possiamo aggiungerne. 

Reviewer: ```suggestion
Abbiamo già visto proprietà integrate del DOM. Ce ne sono molte. Ma tecnicamente nessuna di esse ci limita, e se non ci sono sufficienti, possiamo aggiungerne. 
```

> I nodi del DOM sono regolari oggetti Javascript. Possiamo modificarli. 

Reviewer: ```suggestion
I nodi del DOM sono regolari oggetti JavaScript. Possiamo modificarli. 
```

document.body.sayHi(); // Hello, I'm BODY
```

> Quindi, le proprietà e i metodi del DOM si comportano come regolari oggetti Javascript:

Reviewer: ```suggestion
Quindi, le proprietà e i metodi del DOM si comportano come regolari oggetti JavaScript:
```

> Quindi, le proprietà e i metodi del DOM si comportano come regolari oggetti JavaScript:

Reviewer 2: ```suggestion
Quindi, le proprietà e i metodi del DOM si comportano come normali oggetti JavaScript:
```

- Possono avere qualsiasi valore.
- Sono sensibili alle maiuscole (scrivi `elem.nodeType`, non `elem.NoDeTyPe`).

## Attributi HTML 

> In HTML, i tag possono avere attributi. Quando il browser analizza l'HTML per creare oggetti DOM per i tag, riconosce gli attributi *standart* e crea proprietà del DOM a partire da essi.

Reviewer: ```suggestion
In HTML, i tag possono avere attributi. Quando il browser analizza l'HTML per creare gli oggetti DOM corrispondenti, riconosce gli attributi *standard* e crea proprietà del DOM a partire da essi.
```

> Perciò, quando un elemento ha `id` o un altro attributo *standard*, la proprietà corrispondente viene creata. Ma questo non succede se l'attributo non è standart. 

Reviewer: ```suggestion
Perciò, quando un elemento ha `id` o un altro attributo *standard*, la proprietà corrispondente viene creata. Ma questo non succede se l'attributo non è standard. 
```

Ad esempio:
```html run
<body id="test" something="non-standard">
  <script>
    alert(document.body.id); // test
*!*
    // un attributo non standard non ritorna una proprietà
    alert(document.body.something); // undefined
*/!*
  </script>
</body>
```

> Nota che un attributo standart di un elemento può essere sconosciuto ad un altro. Ad esempio, `"type"` è standard per `<input>` ([HTMLInputElement](https://html.spec.whatwg.org/#htmlinputelement)), ma non per `<body>` ([HTMLBodyElement](https://html.spec.whatwg.org/#htmlbodyelement)). Gli attributi standart sono descritti nella specifica della classe corrispondente dell'elemento.

Reviewer: ```suggestion
Nota che un attributo standard di un elemento può essere sconosciuto ad un altro. Ad esempio, `"type"` è standard per `<input>` ([HTMLInputElement](https://html.spec.whatwg.org/#htmlinputelement)), ma non per `<body>` ([HTMLBodyElement](https://html.spec.whatwg.org/#htmlbodyelement)). Gli attributi standard sono descritti nella specifica della classe corrispondente dell'elemento.
```

Possiamo vederlo qui:
```html run
<body id="body" type="...">
  <input id="input" type="text">
  <script>
    alert(input.type); // text
*!*

>     alert(body.type); // undefined: proprietà del DOM non creata perché non standard

Reviewer 2: ```suggestion
    alert(body.type); // undefined: proprietà del DOM non creata, perché non standard
```

*/!*
  </script>
</body>
```

> Dunque, se un attributo non è standart, non vi sarà una corrispondente proprietà del DOM. C'è un modo per accedere a questi attributi?

Reviewer: ```suggestion
Dunque, se un attributo non è standard, non vi sarà una corrispondente proprietà del DOM. C'è un modo per accedere a questi attributi?
```

Certo. Tutti gli attributi sono accessibili utilizzando i seguenti metodi:

- `elem.hasAttribute(name)` -- controlla l'esistenza.
- `elem.getAttribute(name)` -- ritorna il valore.
- `elem.setAttribute(name, value)` -- imposta un valore.
- `elem.removeAttribute(name)` -- rimuove l'attributo.

Questi metodi operano con quello che è scritto in HTML.

> Si possono altresì leggere tutti gli attributi utilizzando `elem.attributes`: una collection di oggetti che appartengono alla classe integrata [Attr](https://dom.spec.whatwg.org/#attr), con proprietà come`name` e `value`.

Reviewer: ```suggestion
Si possono altresì leggere tutti gli attributi utilizzando `elem.attributes`: una collection di oggetti che appartengono alla classe integrata [Attr](https://dom.spec.whatwg.org/#attr), con proprietà come `name` e `value`.
```

> Segue un esempio di lettura di proprietà non standart:

Reviewer: ```suggestion
Segue un esempio di lettura di proprietà non standard:
```

```html run
<body something="non-standard">
  <script>
*!*
    alert(document.body.getAttribute('something')); // non standard
*/!*
  </script>
</body>
```

Gli attributi HTML hanno le seguenti caratteristiche:

- I loro nomi sono insensibili alle maiuscole (`id` è lo stesso di `ID`). 
- I loro valori sono sempre stringhe.

Ecco un esempio esaustivo di lavoro con gli attributi:

```html run
<body>
  <div id="elem" about="Elephant"></div>

  <script>
    alert( elem.getAttribute('About') ); // (1) 'Elephant', legge

    elem.setAttribute('Test', 123); // (2), scrive

    alert( elem.outerHTML ); // (3), controlla se l'attributo è nell'HTML (sì)

    for (let attr of elem.attributes) { // (4) li elenca tutti
      alert( `${attr.name} = ${attr.value}` );
    }
  </script>
</body>
```

Nota:

> 1. `getAttribute('About')` -- la prima lettera è maiuscola, mentre in HTML è minuscola. Ma non importa: i nomi degli attributi sono insensibili alle maiuscole. 

Reviewer: ```suggestion
1. `getAttribute('About')` -- la prima lettera è maiuscola, mentre in HTML è minuscola. Non ha importanza: i nomi degli attributi sono insensibili alle maiuscole. 
```

2. Possiamo assegnare qualsiasi cosa a un attributo, ma diventerà una stringa. Perciò qui abbiamo `"123"` come valore.
3. Tutti gli attributi, inclusi quelli che abbiamo impostato, sono visibili in `outerHTML`.
4. La collection `attributes` è iterabile e ha tutti gli attributi dell'elemento (standard e non standard) come oggetti con proprietà `name` e `value`.

## Sincronizzazione proprietà-attributo

> Quando un attributo standart viene cambiato, la proprietà corrispondente viene automaticamente aggiornata, e (con qualche eccezione) vice versa.

Reviewer: ```suggestion
Quando un attributo standard viene cambiato, la proprietà corrispondente viene automaticamente aggiornata, e (con qualche eccezione) vice versa.
```

```html run
<input>

<script>
  let input = document.querySelector('input');

  // attributo => proprietà
  input.setAttribute('id', 'id');
  alert(input.id); // id (aggiornato)

  // proprietà => attributo
  input.id = 'newId';
  alert(input.getAttribute('id')); // newId (aggiornato)
</script>
```

Ma ci sono eccezioni; ad esempio `input.value` si sincronizza solo da attributo a proprietà, ma non al contrario:

```html run
<input>

<script>
  let input = document.querySelector('input');

  // attributo => proprietà
  input.setAttribute('value', 'text');
  alert(input.value); // text

*!*
  // NON proprietà => attributo
  input.value = 'newValue';
  alert(input.getAttribute('value')); // text (non aggiornato!)
*/!*
</script>
```

Nell'esempio sopra:
- Cambiare l'attributo `value` aggiorna la proprietà corrispondente.
- Ma la modifica della proprietà non cambia l'attributo. 

Questa caratteristica può essere utile perché le azioni dell'utente potrebbero portare a cambi di `value`, e dopo questi, se vogliamo recuperare il valore "originale" dall'HTML, lo troveremo nell'attributo.

> ## Le proprietà del DOM sono tipate

Reviewer: ```suggestion
## Le proprietà del DOM sono tipizzate
```

Le proprietà del DOM non sono sempre stringhe. Ad esempio, la proprietà `input.checked` (per i checkbox) è un booleano:

```html run
<input id="input" type="checkbox" checked> checkbox

<script>
  alert(input.getAttribute('checked')); // il valore dell'attributo è: stringa vuota
  alert(input.checked); // il valore della proprietà è: true
</script>
```

> Ci sono altri esempi. L'attributo `style` è una stringa, ma la proprietà `style` non la è.

Reviewer: ```suggestion
Ci sono altri esempi. L'attributo `style` è una stringa, ma la proprietà `style` non lo è.
```

</script>
```

> La maggior parte dell proprietà, comunque, sono stringhe.

Reviewer: ```suggestion
La maggior parte delle proprietà, comunque, sono stringhe.
```

> Raramente, anche se il tipo di una proprietà è una stringa, questa potrebbe essere diversa dall'attributo. Ad esempio, la proprietà DOM `href` è sempre una URL *intera*, anche se l'attributo contiene una URL relativa o semplicemente un `#hash`.

Reviewer 2: ```suggestion
Raramente, anche se il tipo di una proprietà è una stringa, questa potrebbe essere diversa dall'attributo. Ad esempio, la proprietà DOM `href` è sempre un URL *completo*, anche se l'attributo contiene un URL relativo o semplicemente un `#hash`.
```

Un esempio:

```html height=30 run
<a id="a" href="#hello">link</a>
<script>
  // attributo
  alert(a.getAttribute('href')); // #hello

  // proprietà
  alert(a.href ); // URL intera nella forma http://site.com/page#hello
</script>
```

Se necessitiamo del valore di `href` o di un qualsiasi altro attributo così come si trova nell'HTML, possiamo utilizzare `getAttribute`.

## Attributi non standard, dataset

> Quando scriviamo HTML, utilizziamo molti attributi standart. E quelli non standart, personalizzati? Prima di tutto, vediamo se sono utili o meno, e per cosa.

Reviewer: ```suggestion
Quando scriviamo HTML, utilizziamo molti attributi standard. E quelli non standard, personalizzati? Prima di tutto, vediamo se sono utili o meno, e per cosa.
```

> Qualche volta gli attributi non standart sono utilizzati per passare dati personalizzati dall'HTML a Javascript, o per "contrassegnare" elementi.

Reviewer: ```suggestion
Qualche volta gli attributi non standard sono utilizzati per passare dati personalizzati dall'HTML a JavaScript, o per "contrassegnare" elementi.
```

</style>

<div class="order" order-state="new">
  Un nuovo ordine.
</div>

<div class="order" order-state="pending">
  Un ordine in attesa.
</div>

<div class="order" order-state="canceled">
  Un ordine cancellato.
</div>
```

Perché un attributo dovrebbe essere preferibile a una classe come `.order-state-new`, `.order-state-pending`, `.order-state-canceled`?

Perché un attributo è più semplice da gestire. Lo stato può cambiare così:

```js
// un po' più semplice del rimuovere/aggiungere una nuova classe
div.setAttribute('order-state', 'canceled');
```

> Ma c'è un potenziale problema con gli attributi personalizzati. Che succederebbe se utilizzassimo un attributo non standart che in seguito verrebbe reso tale, con le sue caratteristiche? Il linguaggio HTML è vivo, cresce, e altri attributi potrebbero apparire per soddisfare le necesssità degli sviluppatori.

Reviewer: ```suggestion
Ma c'è un potenziale problema con gli attributi personalizzati. Che succederebbe se utilizzassimo un attributo non standard che in seguito verrebbe reso tale, con le sue caratteristiche? Il linguaggio HTML è vivo, cresce, e altri attributi potrebbero apparire per soddisfare le necessità degli sviluppatori.
```

> Per evitare conflitti, abbiamo gli attributi [data-*](https://html.spec.whatwg.org/#embedding-custom-non-visible-data-with-the-data-*-attributes).

Reviewer: ```suggestion
Per evitare conflitti, abbiamo a disposizione gli attributi [data-*](https://html.spec.whatwg.org/#embedding-custom-non-visible-data-with-the-data-*-attributes).
```

</style>

<div id="order" class="order" data-order-state="new">
  Un nuovo ordine.
</div>

<script>
  // legge
  alert(order.dataset.orderState); // new

  // modifica
  order.dataset.orderState = "pending"; // (*)
</script>
```

Utilizzare attributi `data-*` è un modo valido e sicuro per passare dati personalizzati. 

Nota, possiamo non solo leggere, ma anche modificare "attributi-data". CSS aggiorna l'aspetto conseguentemente: nell'esempio sopra l'ultima linea `(*)` cambia il colore a blu.

## Riepilogo

- Attributi -- quel che è scritto in HTML.
- Proprietà -- quel che si trova negli oggetti DOM.

Una piccola comparazione:

|            | Proprietà | Attributi  |
|------------|------------|------------|

> |Tipo|Qualsiasi valore, le proprietà standard hanno i tipi descritti descritti nella specifica|Una stringa|

Reviewer: ```suggestion
|Tipo|Qualsiasi valore, le proprietà standard hanno i tipi descritti nella specifica|Una stringa|
```

|Nome|Nome è sensibile alle maiuscole|Nome non è sensibile alle maiuscole|

I metodi per lavorare con gli attributi sono:

- `elem.hasAttribute(name)` -- controlla l'esistenza.
- `elem.getAttribute(name)` -- ritorna il valore.
- `elem.setAttribute(name, value)` -- imposta un valore.
- `elem.removeAttribute(name)` -- rimuove l'attributo.
- `elem.attributes` una collection di tutti gli attributi.

Nella maggior parte delle situazioni è preferibile utilizzare le proprietà del DOM. Dovremmo riferirci agli attributi solo quando le proprietà non sono sufficienti, quando abbiamo bisogno dell'attributo esatto; ad esempio:

> - Abbiamo bisogno di un attributo non standart. Ma se comincia con `data-`, allora dovremmo utilizzare `dataset`.

Reviewer: ```suggestion
- Abbiamo bisogno di un attributo non standard. Ma se comincia con `data-`, allora dovremmo utilizzare `dataset`.
```

> - Dobbiamo leggere un valore così "come scritto" in HTML. Il valore della corrispondente proprietà DOM potrebbe essere diverso; ad esempio la proprietà `href` è sempre una URL intera, e vorremmo il valore "originale".

Reviewer 2: ```suggestion
- Dobbiamo leggere un valore così "come scritto" in HTML. Il valore della corrispondente proprietà DOM potrebbe essere diverso; ad esempio la proprietà `href` è sempre un URL completo, e vorremmo il valore "originale".
```
